src/Management/src/Endpoint/Actuators/HttpExchanges/HttpExchangesDiagnosticObserver.cs

```diff
@@ -0,0 +1,202 @@
+    }
+
+    public HttpExchangesResult GetHttpExchanges()
+    {
+        return new HttpExchangesResult(Queue.ToList());
+    }
+
+    public override void ProcessEvent(string eventName, object? value)
+    {
+        if (eventName != StopEventName)
+        {
+            return;
+        }
+
+        Activity? current = Activity.Current;
+
+        if (current == null)
+        {
+            return;
+        }
+
+        if (value == null)
+        {
+            return;
+        }
+
+        HttpContext? context = GetHttpContextPropertyValue(value);
+
+        if (context != null)
+        {
+            RecordHttpTrace(current, context);
+        }
+    }
+
+    private HttpExchange GetHttpExchange(HttpContext context, TimeSpan duration)
+    {
+        HttpExchangesEndpointOptions options = _optionsMonitor.CurrentValue;
+        string requestUri = GetRequestUri(context.Request, options.AddPathInfo);
+
+        if (options.AddQueryString)
+        {
+            requestUri += context.Request.QueryString.Value;
+        }
+
+        string? remoteAddress = null;
+
+        if (options.AddRemoteAddress)
+        {
+            remoteAddress = GetRemoteAddress(context);
+        }
+
+        Dictionary<string, IList<string?>> requestHeaders = [];
+
+        if (options.AddRequestHeaders)
+        {
+            requestHeaders = GetHeaders(context.Request.Headers, options.RequestHeaders);
+        }
+
+        var traceRequest = new TraceRequest(context.Request.Method, requestUri, requestHeaders, remoteAddress);
+
+        Dictionary<string, IList<string?>> responseHeaders = [];
+
+        if (options.AddResponseHeaders)
+        {
+            responseHeaders = GetHeaders(context.Response.Headers, options.ResponseHeaders);
+        }
+
+        var traceResponse = new TraceResponse(context.Response.StatusCode, responseHeaders);
+
+        string? userName = null;
+
+        if (options.AddUserPrincipal)
+        {
+            userName = GetUserPrincipal(context);
+        }
+
+        TracePrincipal? principal = userName == null ? null : new TracePrincipal(userName);
+
+        string? sessionId = null;
+
+        if (options.AddSessionId)
+        {
+            sessionId = GetSessionId(context);
+        }
+
+        TraceSession? session = sessionId == null ? null : new TraceSession(sessionId);
+
+        long timestamp = GetJavaTime(DateTime.UtcNow.Ticks);
+
+        return new HttpExchange(traceRequest, traceResponse, timestamp, principal, session, duration.TotalMilliseconds);
+    }
+
+    internal static HttpContext? GetHttpContextPropertyValue(object instance)
+    {
+        return GetPropertyOrDefault<HttpContext>(instance, "HttpContext");
+    }
+
+    internal static Dictionary<string, IList<string?>> GetHeaders(IHeaderDictionary headers, IList<string> allowedHeaders)
+    {
+        var result = new Dictionary<string, IList<string?>>();
+
+        foreach (KeyValuePair<string, StringValues> pair in headers)
```

Comment:
Can be deconstructed into named variables with Resharper (note it has a bug that it makes every variable nullable; needs correction afterwards).